lib/matplotlib/category.py

```diff
@@ -13,111 +18,154 @@
+                not isinstance(v, VALID_TYPES)) for v in values):
+            return np.asarray(values, dtype=float)
 
-        for lab, loc in vmap.items():
-            vals[vals == lab] = loc
+        # force an update so it also does type checking
+        axis.units.update(values)
 
-        return vals.astype('float')
+        str2idx = np.vectorize(axis.units._mapping.__getitem__,
+                               otypes=[float])
+
+        mapped_value = str2idx(values)
+        return mapped_value
 
     @staticmethod
     def axisinfo(unit, axis):
-        majloc = StrCategoryLocator(axis.unit_data.locs)
-        majfmt = StrCategoryFormatter(axis.unit_data.seq)
+        """Sets the axis ticks and labels
+        """
+        # locator and formatter take mapping dict because
+        # args need to be pass by reference for updates
+        majloc = StrCategoryLocator(axis.units)
+        majfmt = StrCategoryFormatter(axis.units)
         return units.AxisInfo(majloc=majloc, majfmt=majfmt)
 
     @staticmethod
-    def default_units(data, axis):
-        # the conversion call stack is:
+    def default_units(data=None, axis=None):
+        # the conversion call stack is supposed to be
         # default_units->axis_info->convert
-        if axis.unit_data is None:
-            axis.unit_data = UnitData(data)
+        if axis.units is None:
+            axis.units = UnitData(data)
         else:
-            axis.unit_data.update(data)
-        return None
+            axis.units.update(data)
+        return axis.units
 
 
-class StrCategoryLocator(ticker.FixedLocator):
-    def __init__(self, locs):
-        self.locs = locs
-        self.nbins = None
+class StrCategoryLocator(ticker.Locator):
+    """tick at every integer mapping of the string data"""
+    def __init__(self, units):
+        """
+        Parameters
+        -----------
+        units: dict
+              (string, integer) mapping
+        """
+        self._units = units
 
+    def __call__(self):
+        return list(self._units._mapping.values())
 
-class StrCategoryFormatter(ticker.FixedFormatter):
-    def __init__(self, seq):
-        self.seq = seq
-        self.offset_string = ''
+    def tick_values(self, vmin, vmax):
+        return self()
+
+
+class StrCategoryFormatter(ticker.Formatter):
+    """String representation of the data at every tick"""
+    def __init__(self, units):
+        """
+        Parameters
+        ----------
+        units: dict
+              (string, integer) mapping
+        """
+        self._units = units
+
+    def __call__(self, x, pos=None):
+        if pos is None:
+            return ""
+        r_mapping = {v: to_str(k) for k, v in self._units._mapping.items()}
+        return r_mapping.get(int(x), '')
 
 
 class UnitData(object):
-    # debatable makes sense to special code missing values
-    spdict = {'nan': -1.0, 'inf': -2.0, '-inf': -3.0}
 
-    def __init__(self, data):
+    def __init__(self, data=None):
```

Comment:
Given the way that the mapping is updated below I think you have to not take input (as the counter is just incremented) so it will step on user values.

Reply:
:sheep: I am :sleepy: and can not read.  Carry on this is :+1:

Reply:
I thought we needed that? In this case, it's UnitData(['a', 'b', 'c']). Granted, I dunno if it ever gets called with data, or if it's always an indirect via an update...